streaming/base/format/base/writer.py

```diff
@@ -259,16 +232,18 @@ def _write_index(self) -> None:
         }
         with open(filename, 'w') as out:
             json.dump(obj, out, sort_keys=True)
-        self.upload(basename)
+        self.executor.submit(self.cloud_writer.upload_file, basename)
 
     def finish(self) -> None:
         """Finish writing samples."""
         if self.new_samples:
             self.flush_shard()
             self._reset_cache()
         self._write_index()
+        logger.debug(f'Waiting for all shard files to get uploaded to {self.remote}')
+        self.executor.shutdown(wait=True, cancel_futures=False)
```

Comment:
This feature seems only supported from Python 3.9+. For example, see the [documentation](https://docs.python.org/3.8/library/concurrent.futures.html) for Python 3.8 where `cancel_futures` doesn't exist.

I don't see the project requiring the latest version of Python (since it seems [fine with 3.7+](https://github.com/mosaicml/streaming/blob/main/setup.py#L118)). Is this a bug, then?

Reply:
Reported as an [issue](https://github.com/mosaicml/streaming/issues/192) so we can track the discussion better.

Reply:
Thanks for raising an issue. Responded to the issue.